opensm/osm_opensm.c

```diff
@@ -81,86 +80,300 @@ extern int osm_ucast_nue_setup(struct osm_routing_engine *, osm_opensm_t *);
+
+static void destroy_routing_engines(
+	osm_opensm_t *osm);
+
+static void __free_routing_module(
+	void *p_object, void *context);
+
+/** =========================================================================
+ */
+
+cl_status_t osm_register_external_routing_engine(
+	IN osm_opensm_t *osm,
+	IN OUT external_routing_engine_module_t *module,
+	IN void *context)
 {
-	switch (type) {
-	case OSM_ROUTING_ENGINE_TYPE_NONE:
-		return "none";
-	case OSM_ROUTING_ENGINE_TYPE_MINHOP:
-		return "minhop";
-	case OSM_ROUTING_ENGINE_TYPE_UPDN:
-		return "updn";
-	case OSM_ROUTING_ENGINE_TYPE_DNUP:
-		return "dnup";
-	case OSM_ROUTING_ENGINE_TYPE_FILE:
-		return "file";
-	case OSM_ROUTING_ENGINE_TYPE_FTREE:
-		return "ftree";
-	case OSM_ROUTING_ENGINE_TYPE_LASH:
-		return "lash";
-	case OSM_ROUTING_ENGINE_TYPE_DOR:
-		return "dor";
-	case OSM_ROUTING_ENGINE_TYPE_TORUS_2QOS:
-		return "torus-2QoS";
-	case OSM_ROUTING_ENGINE_TYPE_NUE:
-		return "nue";
-	case OSM_ROUTING_ENGINE_TYPE_DFSSSP:
-		return "dfsssp";
-	case OSM_ROUTING_ENGINE_TYPE_SSSP:
-		return "sssp";
-	default:
-		break;
+	cl_status_t status;
+	routing_engine_module_t *copy = NULL;
+
+	if (!osm || !module)
+		return CL_INVALID_PARAMETER;
+
+	OSM_LOG(&osm->log, OSM_LOG_VERBOSE,
+		"Assign type '%d' to external routing engine with name: \'%s\'\n",
+		last_external_routing_engine_type,
+		module->name);
+
+	copy = (routing_engine_module_t *)
+		malloc(sizeof(routing_engine_module_t));
+	copy->name = strdup(module->name);
+	copy->setup = module->setup;
+	copy->type = last_external_routing_engine_type++;
+	copy->context = context;
+
+	status = register_routing_engine(osm, copy);
+	if (status != CL_SUCCESS) {
+		__free_routing_module(copy, NULL);
 	}
-	return "unknown";
+	return status;
 }
 
-osm_routing_engine_type_t osm_routing_engine_type(IN const char *str)
+
+cl_status_t register_builtin_routing_engine(
+	IN osm_opensm_t *osm,
+	IN OUT builtin_routing_engine_module_t *module)
+{
+	cl_status_t status;
+	routing_engine_module_t * copy;
+
+	if (!osm || !module)
+		return CL_INVALID_PARAMETER;
+
+	copy = (routing_engine_module_t *)
+		malloc(sizeof(routing_engine_module_t));
+	copy->name = strdup(module->name);
+	copy->setup = module->setup;
+	copy->type = module->type;
+	copy->context = NULL;
+
+	status = register_routing_engine(osm, copy);
+	if (status != CL_SUCCESS) {
+		__free_routing_module(copy, NULL);
+	}
+	return status;
+}
+
+cl_status_t register_routing_engine(
+	IN osm_opensm_t *osm,
+	IN OUT routing_engine_module_t *module)
+{
+	cl_status_t status;
+	osm_routing_engine_type_t type;
+	const char *routing_engine_type;
+
+	routing_engine_type =
+		module->type < OSM_ROUTING_ENGINE_TYPE_UNKNOWN ?
+		"Built-in routing engine" : "External routing engine";
+
+	type = osm_routing_engine_type(module->name);
+	if (type != OSM_ROUTING_ENGINE_TYPE_UNKNOWN) {
+		OSM_LOG(&osm->log, OSM_LOG_ERROR,
+			"%s with name \'%s\' was already registered with type: '%d'\n",
```

Comment:
Do you mean this kind of implementation in **osmroutingplugin/src/osmroutingplugin.c**:
```C
	external_routing_engine_module_t plugin_routing_engine_module = {
		 "Built-in routing engine",
		  routing_engine_setup,
		  plugin,
	};

	osm_register_external_routing_engine(
		osm, &plugin_routing_engine_module, plugin);

```

`Jun 25 14:21:04 086457 [F19E9080] 0x02 -> construct: External routing engine 'Built-in routing engine' has been registered with type '13'`

Reply:
No.
if you write in the external plugin implementation:
'
	external_routing_engine_module_t plugin_routing_engine_module = {
		 "minhop", //Or any other Built-in routing engine name. 
		  routing_engine_setup,
		  plugin,
	};
'
You will receive an error state 
"External routing engine with name 'minhop' was already registered with type: 1''
instead of
"Built-in routing engine with name 'minhop' was already registered with type: 1''

Reply:
I made the error messages more explicit in commit 48ba582